tools/testing/test_selections.py

```diff
@@ -1,34 +1,51 @@
 import os
 import subprocess
 
-from typing import Dict, List, Tuple
+from typing import Callable, Dict, List, Optional, Tuple
 
 from tools.stats.import_test_stats import get_disabled_tests, get_slow_tests
 
 
 def calculate_shards(
-    num_shards: int, tests: List[str], job_times: Dict[str, float]
+    num_shards: int,
+    tests: List[str],
+    job_times: Dict[str, float],
+    must_serial: Optional[Callable[[str], bool]] = None,
 ) -> List[Tuple[float, List[str]]]:
-    filtered_job_times: Dict[str, float] = {}
+    must_serial = must_serial if must_serial is not None else lambda x: True
+
+    filtered_job_times: Dict[str, float] = dict()
     unknown_jobs: List[str] = []
     for test in tests:
         if test in job_times:
             filtered_job_times[test] = job_times[test]
         else:
             unknown_jobs.append(test)
 
-    # The following attempts to implement a partition approximation greedy algorithm
-    # See more at https://en.wikipedia.org/wiki/Greedy_number_partitioning
     sorted_jobs = sorted(
         filtered_job_times, key=lambda j: filtered_job_times[j], reverse=True
     )
     sharded_jobs: List[Tuple[float, List[str]]] = [(0.0, []) for _ in range(num_shards)]
-    for job in sorted_jobs:
-        min_shard_index = sorted(range(num_shards), key=lambda i: sharded_jobs[i][0])[0]
+
+    serial = [x for x in sorted_jobs if must_serial(x)]
+    parallel = [x for x in sorted_jobs if x not in serial]
+
+    for i in range(0, len(serial)):
+        min_shard_index = sorted(range(num_shards), key=lambda j: sharded_jobs[j][0])[0]
+        curr_shard_time, curr_shard_jobs = sharded_jobs[min_shard_index]
+        curr_shard_jobs.append(serial[i])
+        sharded_jobs[min_shard_index] = (
+            curr_shard_time + filtered_job_times[serial[i]],
+            curr_shard_jobs,
+        )
+
+    # Not the best idea, but attempt to mask the long jobs with other long jobs
+    for i in range(0, len(parallel), 3):
```

Comment:
I guess 3 here is the same as the NUM_PROCS constant, so we need to move it here and import it into run_test?

Reply:
Hmm, I think you mean to use `for i in range(0, len(parallel), NUM_PROCS)` here and in line 49 `curr_shard_jobs.extend(parallel[i : i + NUM_PROCS])`

Reply:
whoops, moved the variable and the promptly forgot to use it, fixed now